core/src/main/java/org/elasticsearch/search/aggregations/bucket/range/ip/IpRangeAggregatorBuilder.java

```diff
@@ -0,0 +1,328 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.search.aggregations.bucket.range.ip;
+
+import java.io.IOException;
+import java.net.InetAddress;
+import java.net.UnknownHostException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+
+import org.apache.lucene.document.InetAddressPoint;
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.common.ParseField;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.network.InetAddresses;
+import org.elasticsearch.common.xcontent.ToXContent;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.search.aggregations.AggregatorFactories.Builder;
+import org.elasticsearch.search.aggregations.AggregatorFactory;
+import org.elasticsearch.search.aggregations.InternalAggregation;
+import org.elasticsearch.search.aggregations.bucket.range.BinaryRangeAggregator;
+import org.elasticsearch.search.aggregations.bucket.range.BinaryRangeAggregatorFactory;
+import org.elasticsearch.search.aggregations.bucket.range.RangeAggregator;
+import org.elasticsearch.search.aggregations.support.AggregationContext;
+import org.elasticsearch.search.aggregations.support.ValueType;
+import org.elasticsearch.search.aggregations.support.ValuesSource;
+import org.elasticsearch.search.aggregations.support.ValuesSourceAggregatorBuilder;
+import org.elasticsearch.search.aggregations.support.ValuesSourceAggregatorFactory;
+import org.elasticsearch.search.aggregations.support.ValuesSourceConfig;
+import org.elasticsearch.search.aggregations.support.ValuesSourceType;
+
+
+public final class IpRangeAggregatorBuilder
+        extends ValuesSourceAggregatorBuilder<ValuesSource.Bytes, IpRangeAggregatorBuilder> {
+    private static final String NAME = "ip_range";
+    public static final ParseField AGGREGATION_NAME_FIELD = new ParseField(NAME);
+    private static final InternalAggregation.Type TYPE = new InternalAggregation.Type(NAME);
+
+    public static class Range implements ToXContent {
+        private final String key;
+        private final String from;
+        private final String to;
+
+        Range(String key, String from, String to) {
+            if (from != null) {
+                InetAddresses.forString(from);
+            }
+            if (to != null) {
+                InetAddresses.forString(to);
+            }
+            this.key = key;
+            this.from = from;
+            this.to = to;
+        }
+
+        Range(String key, String mask) {
+            String[] splits = mask.split("/");
+            if (splits.length != 2) {
+                throw new IllegalArgumentException("Expected [ip/prefix_length] but got [" + mask + "], which contains more than one [/]");
```

Comment:
Maybe this message could be improved as it looks this exception would be thrown if the mask contains no `/` characters as well?